sky/spot/recovery_strategy.py

```diff
@@ -211,37 +196,47 @@ def recover(self) -> float:
             logger.info('Ignoring the job cancellation failure; the spot '
                         'cluster is likely completely stopped. Recovering.')
 
-        # Add region constraint to the task, to retry on the same region
-        # first.
-        task = self.dag.tasks[0]
-        resources = list(task.resources)[0]
-        original_resources = resources
-
-        launched_cloud = handle.launched_resources.cloud
-        launched_region = handle.launched_resources.region
-        new_resources = resources.copy(cloud=launched_cloud,
-                                       region=launched_region)
-        task.set_resources({new_resources})
-        # Not using self.launch to avoid the retry until up logic.
-        launched_time = self._launch(raise_on_failure=False)
-        # Restore the original dag, i.e. reset the region constraint.
-        task.set_resources({original_resources})
-        if launched_time is not None:
-            return launched_time
+        while True:
+            # Add region constraint to the task, to retry on the same region
```

Comment:
Actually, what's the reason for retrying in the same region first?

Reply:
I am trying to make sure the logic is that: we try our best to launch the cluster in the same region due to the data locality and then try on other regions. By retrying all those two steps, we preserve the ratio of the retrying for the same region and all the regions.

Reply:
I think it's a fine rationale. We need empirical data to decide on a good strategy (for future). E.g., it is possible the current region that just preempted a cluster likely doesn't get a free resource immediately after.